src/Relation/Nullary/Finite/Setoid.agda

```diff
@@ -0,0 +1,56 @@
+------------------------------------------------------------------------
+-- The Agda standard library
+--
+-- Notions of finiteness for setoids
+------------------------------------------------------------------------
+
+{-# OPTIONS --safe --cubical-compatible #-}
+
+module Relation.Nullary.Finite.Setoid where
+
+open import Data.Fin.Base using (Fin)
+open import Data.Nat.Base using (ℕ)
+open import Function.Bundles using (Injection; Surjection; Inverse)
+open import Level using (Level; _⊔_) renaming (suc to lsuc)
+open import Relation.Binary.Bundles using (Setoid)
+open import Relation.Binary.PropositionalEquality.Properties using (setoid)
+
+private
+  variable
+    c ℓ c′ ℓ′ : Level
+
+record IsStronglyFinite (X : Setoid c ℓ) (n : ℕ) : Set (c ⊔ ℓ) where
+   field
+     inv : Inverse X (setoid (Fin n))
+
+record StronglyFinite (X : Setoid c ℓ) (n : ℕ) : Set (c ⊔ ℓ) where
+   field
+     isFinite : IsStronglyFinite X n
+
+record Subfinite (X : Setoid c ℓ) : Set (c ⊔ ℓ) where
+  field
+    size : ℕ
+    inj : Injection X (setoid (Fin size))
+
+record FinitelyEnumerable (X : Setoid c ℓ) : Set (c ⊔ ℓ) where
+  field
+    size : ℕ
+    srj : Surjection (setoid (Fin size)) X
+
+record InjectsIntoFinitelyEnumerable (X : Setoid c ℓ) c′ ℓ′
+       : Set (c ⊔ ℓ ⊔ lsuc (c′ ⊔ ℓ′)) where
+  field
+    Apex : Setoid c′ ℓ′
+    finitelyEnumerable : FinitelyEnumerable Apex
+    inj : Injection X Apex
+
+  open FinitelyEnumerable finitelyEnumerable public
+
+record SurjectionFromFinitelyEnumerable (X : Setoid c ℓ) c′ ℓ′
+       : Set (c ⊔ ℓ ⊔ lsuc (c′ ⊔ ℓ′)) where
+  field
+    Apex : Setoid c′ ℓ′
+    subfinite : Subfinite Apex
+    srj : Surjection Apex X
```

Comment:
See above. I'd prefer a 'better' name, even though I appreciate it's useful to have a field name that is clearly 'standardised apart' from any potential clashes...

... ob: which, will `Apex` ever be mentioned explicitly in use of these records, or could it be declared implicit, as it's governed by the more-informative subsequent field name in each case?

Reply:
I suppose being able to answer the latter question is why PRs normally come with proofs. They'd give some idea as to what the relevant idioms are.

For what it's worth, I would guess that `Apex` often does have to be given and referred to explicitly. At least in basic cases, I expect that a lot of the work will go into finding the right setoid, and then backward-chaining all of the required maps.

Reply:
I agree that PRs that introduce new ideas should not be so minimalistic. (My preference for small PRs was pushed too far here.)